### DISCUSSION TITLE
First and last tick labels are omitted

### DISCUSSION THREAD

**Discussion by alpriest** [patrykandpatrick/vico#431] [Ideas]
**Describe the bug**
When the number of data points is over a certain threshold and when scrolling is disabled, the initial and final chart tick labels are omitted and instead replaced with ... 

**Module**

- [x] `compose`
- [ ] `views`
- [ ] `compose-m2`
- [ ] `compose-m3`
- [ ] `core` (select if none of the above apply)

**To reproduce**
```
@Preview
@Composable
fun StatsGraphViewDemo() {
    val values = entryModelOf((1..31).map { FloatEntry(it.toFloat(), 5f) })

    Column(modifier = Modifier.fillMaxWidth()) {
        Chart(
            chart = columnChart(),
            model = values,
            startAxis = startAxis(),
            bottomAxis = bottomAxis(
                tickPosition = HorizontalAxis.TickPosition.Center(offset = 0, spacing = 3),
            ),
            chartScrollSpec = rememberChartScrollSpec(isScrollEnabled = false)
        )
    }
}
```

**Expected behavior**
I would expect the initial and final values to be displayed, if necessary padding the graph by sufficient pixels to display (in the event of the final value).

**Screenshots**
<img width="555" alt="Screenshot 2023-05-21 at 17 17 29" src="https://github.com/patrykandpatrick/vico/assets/221241/a98daa51-6071-48a7-90e3-bb05d342a713">

**Android versions**
Tested on API level 33.

**Additional context**
A workaround is to specify the `offset = 1`, but i'd really like the graph to have the first value labelled.

Omitting `offset = 0` from the above example omits the first tick entirely despite the default parameter appearing to be 0.

**Comment by patrickmichalik** [comment]
Hello! Thank you for your input. We’ll be looking into improving how this is handled. It’s worth noting that `axisLabelComponent` has some default padding. Setting the padding to zero might make the first label visible, as it will require less horizontal space.

**Comment by alpriest** [comment]
Thanks for mentioning the padding, i've overridden it and am able to get the first element visible. The last is sometimes omitted still but it's an improvement.

**Comment by patrickmichalik** [comment]
[Vico 1.7.0](https://github.com/patrykandpatrick/vico/releases/tag/v1.7.0) introduced `HorizontalLayout`. By using the `HorizontalLayout.fullWidth` function’s `startPadding` and `endPadding` parameters, you can add enough horizontal padding for the labels to be visible. The padding is scaled with the chart, so this might take some playing around with, but you should be able to get it to work. Feel free to let me know if you face any issues. As per your suggestion, we plan on adding a feature that automatically adds the required amount of padding.

**Comment by agustinsivoplas** [comment]
@patrickmichalik I tried the solution to add start/end padding (12 dp) to make the labels visible, the thing is now the guidelines overflow the chart.

![Screenshot_1697060032](https://github.com/patrykandpatrick/vico/assets/1648539/cdc22b26-0f7d-4113-a8d6-8218258692bc)
![Screenshot_1697060090](https://github.com/patrykandpatrick/vico/assets/1648539/567dafcf-0117-436d-b7fe-e56ca171b586)

So I decided to add the same padding to the guidelines (12 dp too) but not luck.

![Screenshot_1697061243](https://github.com/patrykandpatrick/vico/assets/1648539/ab977715-7ecb-4749-8db3-27477d5464d6)
![Screenshot_1697061352](https://github.com/patrykandpatrick/vico/assets/1648539/ac8aecc4-8f60-46c3-a1ac-ebe8df9f65d7)

So my other test was add (start/end padding = 12.dp and guidelines horizontal padding = 18.dp)

The mobile version looks good, but the tablet version still having issues:

![Screenshot_1697061487](https://github.com/patrykandpatrick/vico/assets/1648539/861cd57c-b326-464c-ac45-b1e8f81cd043)
![Screenshot_1697061491](https://github.com/patrykandpatrick/vico/assets/1648539/81ca306a-42a5-4459-9fbd-d2d0b55b2587)

If I setup (start/end padding = 12.dp and guidelines horizontal padding = 26.dp)

The tablet version looks good but now the mobile version is broken.

![Screenshot_1697061726](https://github.com/patrykandpatrick/vico/assets/1648539/99ffc54e-889c-4455-93c6-d6b9ad66bd8a)
![Screenshot_1697061728](https://github.com/patrykandpatrick/vico/assets/1648539/6d4ddab3-a4ee-4f03-983a-316445d22eca)


Also the distance between Mild/Mod/Sev and Sun label started to grow a lot of.


Even if I play with the paddings I am not able to get good results in mobile/tablet at same time.

Any other suggestion?

**Reply by patrickmichalik** [reply]
Hello! The reason why this approach doesn’t consistently produce the desired result is that the padding is scaled with the chart, but the width of the guidelines isn’t. When scrolling is disabled or `AutoScaleUp.Full` is used, the chart’s scale is determined automatically. The zoom factor isn’t necessarily 1, and it may differ across devices. [Vico 1.13.0 Alpha 7](https://github.com/patrykandpatrick/vico/releases/tag/v1.13.0-alpha.7), which we’ve just released, enables you to add unscalable padding via `HorizontalLayout.FullWidth`. Please try using the solution you’ve described, but making the 12 dp of padding unscalable (via `unscalable*PaddingDp`). That should give you the desired result across all devices, in terms of both the guidelines and the labels. If you need an example, feel free to let me know. (@alpriest, because your chart is unscrollable, the ability to add unscalable padding may be of interest to you too.)

**Comment by agustinsivoplas** [comment]
> Hello! The reason why this approach doesn’t consistently produce the desired result is that the padding is scaled with the chart, but the width of the guidelines isn’t. When scrolling is disabled or `AutoScaleUp.Full` is used, the chart’s scale is determined automatically. The zoom factor isn’t necessarily 1, and it may differ across devices. [Vico 1.13.0 Alpha 7](https://github.com/patrykandpatrick/vico/releases/tag/v1.13.0-alpha.7), which we’ve just released, enables you to add unscalable padding via `HorizontalLayout.FullWidth`. Please try using the solution you’ve described, but making the 12 dp of padding unscalable (via `unscalable*PaddingDp`). That should give you the desired result across all devices, in terms of both the guidelines and the labels. If you need an example, feel free to let me know. (@alpriest, because your chart is unscrollable, the ability to add unscalable padding may be of interest to you too.)

Thanks for the quick answer and solution. Unfortunately I still having issues at the end of the chart (check images and code maybe I am doing something wrong) and also I have another issue with labels. If you see my images we allow the user to create charts for last 30 days /6 months, that means that we need to add 30 or 180 points to the X-Axis, so are a lot of labels. The solution was use the valueFormatter to limit the number of labels but still not working


**(Daily chart) end of chart issuue**

![Screenshot_1697465283](https://github.com/patrykandpatrick/vico/assets/1648539/d318b744-542a-4e5a-b4c6-dbb8c6755cbb)

**(30 days and 6 months) labels and end of chart issue**

![Screenshot_1697465291](https://github.com/patrykandpatrick/vico/assets/1648539/9fcf41b8-a3a0-4050-b9ec-7458b025d38d)
![Screenshot_1697465295](https://github.com/patrykandpatrick/vico/assets/1648539/372e9133-178f-4f08-916b-5fb0df33dc56)


```
        Chart(
            modifier = modifier.height(SymptomTrackerChartHeight),
            isZoomEnabled = false,
            chart = lineChart(
                lines = listOf(
                    lineSpec(
                        lineColor = ColorTealLight,
                        lineBackgroundShader = verticalGradient(arrayOf(ColorTealLight.copy(0.5f), ColorTealLight.copy(alpha = 0.1f))),
                        pointConnector = DefaultPointConnector(cubicStrength = 0f)
                    )
                ),
                axisValuesOverrider = AxisValuesOverrider.fixed(
                    maxY = SymptomSeverity.entries.size.toFloat()
                ),
            ),
            model = chartItem.entryModel,
            horizontalLayout = HorizontalLayout.fullWidth(
                unscalableStartPadding = ChartLabelGuidelineHorizontalMargin,
                unscalableEndPadding = ChartLabelGuidelineHorizontalMargin
            ),
            startAxis = rememberStartAxis(
                label = axisLabelComponent(
                    color = MaterialTheme.colorScheme.onPrimaryContainer,
                    textSize = TextSizeSmall,
                    typeface = typeface ?: Typeface.DEFAULT,
                    horizontalPadding = 0.dp,
                    horizontalMargin = 0.dp
                ),
                guideline = axisLineComponent(
                    thickness = ChartGuidelineThickness,
                    margins = dimensionsOf(horizontal = ChartLabelGuidelineHorizontalMargin)
                ),
                tick = null,
                axis = null,
                itemPlacer = remember { AxisItemPlacer.Vertical.default(maxItemCount = SymptomSeverity.entries.size + 1) },
                valueFormatter = { y, _ ->
                    if (y > 0) {
                        context.getString(SymptomSeverity.entries[y.toInt() - 1].displayChartResId)
                    } else {
                        ""
                    }
                }
            ),
            bottomAxis = rememberBottomAxis(
                axis = axisLineComponent(
                    thickness = ChartGuidelineThickness,
                    margins = dimensionsOf(horizontal = ChartLabelGuidelineHorizontalMargin)
                ),
                guideline = null,
                tick = null,
                label = axisLabelComponent(
                    color = MaterialTheme.colorScheme.onPrimaryContainer,
                    textSize = TextSizeSmall,
                    typeface = typeface ?: Typeface.DEFAULT,
                    horizontalPadding = 0.dp,
                    horizontalMargin = 0.dp
                ),
                valueFormatter = { x, _ ->
                    when (uiState.symptomsTrackerTimeSelector) {
                        SymptomsTrackerTimeSelector.LAST_7_DAYS -> LocalDate.ofEpochDay(x.toLong()).formatDayShort()
                        SymptomsTrackerTimeSelector.LAST_30_DAYS -> {
                            if (x.toInt() % 5 == 0) {
                                LocalDate.ofEpochDay(x.toLong()).formatDayAndMonthShort()
                            } else {
                                ""
                            }
                        }

                        SymptomsTrackerTimeSelector.LAST_6_MONTHS -> {
                            if (x.toInt() % 30 == 0) {
                                LocalDate.ofEpochDay(x.toLong()).formatMonthShort()
                            } else {
                                ""
                            }
                        }
                    }
                }
            ),
            chartScrollSpec = rememberChartScrollSpec(isScrollEnabled = false)
        )
```

In tablets I have a similar issue but is consistency with mobile issues.

![Screenshot_1697465603](https://github.com/patrykandpatrick/vico/assets/1648539/b0f2c072-5f41-41a5-8c10-c47f5c19381a)
![Screenshot_1697465608](https://github.com/patrykandpatrick/vico/assets/1648539/90dfa12b-75a6-4eb5-a07a-bc32c27735a6)

**Reply by patrickmichalik** [reply]
The issue where the guidelines stick out past the end of the chart is the result of a bug, for which we apologize. [Vico 1.13.0 Alpha 8](https://github.com/patrykandpatrick/vico/releases/tag/v1.13.0-alpha.8), which we’ve just released, should fix the problem. Please try updating to this version.

Regarding the `HorizontalAxis` labels, we don’t recommend the approach described. An empty label and no label aren’t entirely equivalent. Please try using `AxisItemPlacer.Horizontal.default`, which has a `spacing` parameter, instead.

```kt
rememberBottomAxis(
    valueFormatter = { x, _ ->
        LocalDate.ofEpochDay(x.toLong()).run {
            when (uiState.symptomsTrackerTimeSelector) {
                SymptomsTrackerTimeSelector.LAST_7_DAYS -> formatDayShort()
                SymptomsTrackerTimeSelector.LAST_30_DAYS -> formatDayAndMonthShort()
                SymptomsTrackerTimeSelector.LAST_6_MONTHS -> formatMonthShort()
            }
        }
    },
    itemPlacer = AxisItemPlacer.Horizontal.default(
        spacing = when (uiState.symptomsTrackerTimeSelector) {
            SymptomsTrackerTimeSelector.LAST_7_DAYS -> 1
            SymptomsTrackerTimeSelector.LAST_30_DAYS -> 5
            SymptomsTrackerTimeSelector.LAST_6_MONTHS -> 30
        },
    ),
)
```

**Comment by agustinsivoplas** [comment]
The temporal solution that I found for now was disable the X-Axis label and create a fake bottom axis. Is not ideal but we have time until launch to production so I hope we can fix using the library :)

Final (temporary) code is this:

```
        Chart(
            modifier = modifier.height(SymptomTrackerChartHeight),
            isZoomEnabled = false,
            chart = lineChart(
                lines = listOf(
                    lineSpec(
                        lineColor = ColorTealLight,
                        lineBackgroundShader = verticalGradient(arrayOf(ColorTealLight.copy(0.5f), ColorTealLight.copy(alpha = 0.1f))),
                        pointConnector = DefaultPointConnector(cubicStrength = 0f)
                    )
                ),
                axisValuesOverrider = AxisValuesOverrider.fixed(
                    maxY = SymptomSeverity.entries.size.toFloat()
                )
            ),
            model = chartItem.entryModel,
            horizontalLayout = HorizontalLayout.fullWidth(),
            startAxis = rememberStartAxis(
                label = axisLabelComponent(
                    color = MaterialTheme.colorScheme.onPrimaryContainer,
                    textSize = TextSizeSmall,
                    typeface = typeface ?: Typeface.DEFAULT
                ),
                guideline = axisLineComponent(
                    thickness = ChartGuidelineThickness
                ),
                tick = null,
                axis = null,
                itemPlacer = remember { AxisItemPlacer.Vertical.default(maxItemCount = SymptomSeverity.entries.size + 1) },
                valueFormatter = { y, _ ->
                    if (y > 0) {
                        context.getString(SymptomSeverity.entries[y.toInt() - 1].displayChartResId)
                    } else {
                        ""
                    }
                }
            ),
            bottomAxis = rememberBottomAxis(
                axis = axisLineComponent(
                    thickness = ChartGuidelineThickness
                ),
                guideline = null,
                tick = null,
                label = null
            ),
            chartScrollSpec = rememberChartScrollSpec(isScrollEnabled = false)
        )

        //Fake X-axis since the library does not support limit the labels at bottom yet.
        Row(
            modifier = modifier
                .padding(start = DefaultPaddingExtraLargeMedium)
                .fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween
        ) {
            chartItem.entryModel.entries.forEach {
                it.forEach { entry ->
                    when (uiState.symptomsTrackerTimeSelector) {
                        SymptomsTrackerTimeSelector.LAST_7_DAYS -> {
                            BodySmallSemiBold(text = LocalDate.ofEpochDay(entry.x.toLong()).formatDayShort())
                        }

                        SymptomsTrackerTimeSelector.LAST_30_DAYS -> {
                            if (entry.x.toInt() % 5 == 0) {
                                BodySmallSemiBold(text = LocalDate.ofEpochDay(entry.x.toLong()).formatDayAndMonthShort())
                            }
                        }

                        SymptomsTrackerTimeSelector.LAST_6_MONTHS -> {
                            if (entry.x.toInt() % 30 == 0) {
                                BodySmallSemiBold(text = LocalDate.ofEpochDay(entry.x.toLong()).formatMonthShort())
                            }
                        }
                    }
                }
            }
        } 
```


![Screenshot_1697466027](https://github.com/patrykandpatrick/vico/assets/1648539/005eda64-5321-45b9-81f6-d5c7e2c53821)
![Screenshot_1697466031](https://github.com/patrykandpatrick/vico/assets/1648539/26b717cb-4d35-41af-b39a-96d2720affad)
![Screenshot_1697466033](https://github.com/patrykandpatrick/vico/assets/1648539/60ad9c32-7349-4237-aeab-2217c5b5778a)

**Reply by patrickmichalik** [reply]
While this approach certainly works, Vico does let you customize the spacing of a `HorizontalAxis`’s labels, as I’ve described in [my previous comment](https://github.com/patrykandpatrick/vico/discussions/431#discussioncomment-7297160). I suggest you try out that solution—it should be simpler and more robust. If you face any problems, feel free to let me know.

**Reply by agustinsivoplas** [reply]
Many thanks! Works like a charm!!  I had to adjust some parameters but works

**Reply by patrickmichalik** [reply]
I’m glad to hear that, @agustinsivoplas!

**Comment by patrickmichalik** [comment]
I’m happy to report that [Vico 1.13.0 Alpha 9](https://github.com/patrykandpatrick/vico/releases/tag/v1.13.0-alpha.9) introduces the feature requested here. Please see the release notes for details. Thanks again for the suggestion, @alpriest! @agustinsivoplas, you can now remove the 12 dp of padding and have Vico automatically make room for the leftmost and rightmost labels. Please note that `TextComponent`s created via `axisLabelComponent` have some horizontal padding by default. You might want to set `horizontalPadding` to `0.dp`.

<img src="https://github.com/patrykandpatrick/vico/assets/120058021/5ec0df58-7d59-4d0d-b8c1-b1b04d059194" width="360" />

**Reply by alpriest** [reply]
Thanks, just upgraded and looks great.

**Reply by patrickmichalik** [reply]
I’m glad to hear that!

**Comment by patrickmichalik** [comment]
I’m a bit late here, but to stick to the usual process, I wanted to follow up and mention that [Vico 1.13.0](https://github.com/patrykandpatrick/vico/releases/tag/v1.13.0), a stable release, is now available.
